src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,84 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
-/**
- * Make this class immutable. See requirements in the README file
- */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+    
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = new ArrayList<>();
+        for (Wheel wheel : wheels) {
+            this.wheels.add(wheel.clone());
+        }
+        this.engine = cloneEngine(engine);
+    }
+    
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, cloneWheelsList(), cloneEngine(engine));
+    }
+    
+    public Car changeColor(String newColor) {
+        return new Car(year, newColor, cloneWheelsList(), cloneEngine(engine));
+    }
+    
+    public Car addWheel(Wheel newWheel) {
+        List<Wheel> newWheels = cloneWheelsList();
+        newWheels.add(newWheel.clone());
+        return new Car(year, color, newWheels, cloneEngine(engine));
```

Comment:
```suggestion
       Car car = new Car(year, color, wheels, engine)
       car.wheels.add(newWheel.clone());
        return car;
```
don't make double clonning (clone + inside of constructor